File: src/main/java/com/gemini/krakenbot/model/PortfolioSnapshot.java
package com.gemini.krakenbot.model;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.List;
import java.util.Map;

@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
public class PortfolioSnapshot {
    private Instant timestamp;
    private BigDecimal totalValueUSD;
    private Map<String, AssetSnapshot> assets;
    private List<String> actions;
    private BigDecimal drawdownPercent;
    private BigDecimal fiatDeploymentPercent;
    private BigDecimal effectiveUsdTargetPercent;

    @Getter
    @Setter
    @AllArgsConstructor
    @NoArgsConstructor
    public static class AssetSnapshot {
        private String symbol;
        private BigDecimal balance;
        private BigDecimal price;
        private BigDecimal valueUSD;
        private BigDecimal targetPercent;
        private BigDecimal currentPercent;
        private BigDecimal deviationPercent;
        private BigDecimal deviationUSD;
    }
}

File: src/main/kotlin/com/gemini/krakenbot/model/PortfolioSnapshot.kt
package com.gemini.krakenbot.model

import java.math.BigDecimal
import java.time.Instant

data class PortfolioSnapshot(
    val timestamp: Instant,
    val totalValueUSD: BigDecimal,
    val assets: Map<String, AssetSnapshot>,
    val actions: List<String>,
    val drawdownPercent: BigDecimal,
    val fiatDeploymentPercent: BigDecimal,
    val effectiveUsdTargetPercent: BigDecimal
) {
    data class AssetSnapshot(
        val symbol: Asset,
        val balance: BigDecimal,
        val price: BigDecimal,
        val valueUSD: BigDecimal,
        val targetPercent: BigDecimal,
        val currentPercent: BigDecimal,
        val deviationPercent: BigDecimal,
        val deviationUSD: BigDecimal
    ) {
        companion object {
            operator fun invoke(
                symbol: String,
                balance: BigDecimal,
                price: BigDecimal,
                valueUSD: BigDecimal,
                targetPercent: BigDecimal,
                currentPercent: BigDecimal,
                deviationPercent: BigDecimal,
                deviationUSD: BigDecimal
            ): AssetSnapshot = AssetSnapshot(
                Asset(symbol),
                balance,
                price,
                valueUSD,
                targetPercent,
                currentPercent,
                deviationPercent,
                deviationUSD
            )
        }
    }
}
